src/Xamarin.Android.Build.Tasks/Tasks/GenerateJavaStubs.cs

```diff
@@ -314,17 +318,60 @@ void WriteTypeMappings (List<TypeDefinition> types)
 			using (var gen = UseSharedRuntime
 				? new TypeNameMapGenerator (types, Log.LogDebugMessage)
 			        : new TypeNameMapGenerator (ResolvedAssemblies.Select (p => p.ItemSpec), Log.LogDebugMessage)) {
-				UpdateWhenChanged (Path.Combine (OutputDirectory, "typemap.jm"), gen.WriteJavaToManaged);
-				UpdateWhenChanged (Path.Combine (OutputDirectory, "typemap.mj"), gen.WriteManagedToJava);
+				UpdateWhenChanged (Path.Combine (OutputDirectory, "typemap.jm"), "jm", gen.WriteJavaToManaged);
+				UpdateWhenChanged (Path.Combine (OutputDirectory, "typemap.mj"), "mj", gen.WriteManagedToJava);
 			}
 		}
 
-		void UpdateWhenChanged (string path, Action<Stream> generator)
+		void UpdateWhenChanged (string path, string type, Action<Stream> generator)
 		{
 			using (var stream = new MemoryStream ()) {
 				generator (stream);
 				MonoAndroidHelper.CopyIfStreamChanged (stream, path);
 			}
+
+			string dataFilePath = $"{path}.inc";
+			using (var stream = new NativeAssemblyDataStream ()) {
```

Comment:
We should only use the native library support when `$(EmbedAssembliesIntoApk)` is True.  When `$(EmbedAssembliesIntoApk)` is False -- the default for `Debug` config builds, but *not* `Release` config builds -- the intention is that we fast deploy assemblies onto the device.  Eventually, we'll want to fast deploy the typemap files as well, in which case we don't want the native library version.

Reply:
We always have to use the native library because our runtime is linked against it, so we always need `libxamarin-app.so` in the APK. I think what we need is to skip the generation of the typemap `.mj` and `.jm` files for `Release` builds instead.

Reply:
> I think what we need is to skip the generation of the typemap `.mj` and `.jm` files for `Release` builds instead.

Yes: we should skip generation of the typemap files when `$(EmbedAssembliesIntoApk)` is True; `$(EmbedAssembliesIntoApk)`==False means fast deployment, which in turn implies an (eventual) intent to have *multiple sets* of typemap files which are separately deployed to the device.

When `$(EmbedAssembliesIntoApk)`==True, there shouldn't be a need for separate installation of typemap files, which in turn means we shouldn't need to create them.

Reply:
Yeah, if `EmbedAssemblies` is false, do we need the assembly files?

If that was the case, this change would have basically no impact on Debug/Fast Deployment builds, which would be great!

Reply:
After some discussion on Slack, I think we could actually skip generation of the typemap files in most cases: https://xamarinhq.slack.com/archives/C03CEGRUW/p1555340725186000

This would offset the time taken for generating the assembly files.